File: os-algorithms/DisablingInterrupts.java
public class DisablingInterrupts {
    private volatile boolean interruptsEnabled = true;

    public void enterCriticalSection() {
        interruptsEnabled = false;
    }

    public void exitCriticalSection() {
        interruptsEnabled = true;
    }

    public boolean areInterruptsEnabled() {
        return interruptsEnabled;
    }

    public static void main(String[] args) {
        DisablingInterrupts di = new DisablingInterrupts();

        System.out.println("Interrupts enabled? " + di.areInterruptsEnabled());

        di.enterCriticalSection();
        try {
            System.out.println("Inside critical section. Interrupts enabled? " + di.areInterruptsEnabled());
            // Simple simulation task
            int a = 2, b = 3;
            int result = a * b;
            System.out.println("Result of " + a + " * " + b + " is: " + result);
        } finally {
            di.exitCriticalSection();
        }
        
        System.out.println("Interrupts enabled? " + di.areInterruptsEnabled());
    }
}

File: os-algorithms/disabling_interrupts.cpp
#include <iostream>

void disable_interrupts() {
    asm volatile ("cli" ::: "memory");
}

void enable_interrupts() {
    asm volatile ("sti" ::: "memory");
}

void performCriticalOperation() {
    std::cout << "Inside critical section." << std::endl;
    // Simulate some work in the critical section.
    for (volatile int i = 0; i < 1000000; ++i) {
        // ... performing critical work ...
    }
    std::cout << "Exiting critical section." << std::endl;
}

int main() {
    std::cout << "Starting critical operation with interrupts disabled." << std::endl;
   
    disable_interrupts();
    
    performCriticalOperation();
    
    enable_interrupts();
    
    std::cout << "Critical operation finished, interrupts enabled again." << std::endl;
    return 0;
}
